### PR TITLE
Add utility to send and receive frames over the network

### PR DESCRIPTION
**Pull request by adutra**
This PR also:

1. Exposes the `DecodeHeader`, `DecodeBody` and `DecompressBody` methods of `frame.Codec`. 
2. It also introduces a new struct `RawHeader` and a new `DiscardBody` method. These changes aim to make it possible to only decode the header of a received frame.
3. Fixes lz4 compression, including compression for empty messages.

### CODE REVIEW

File: cassandraprotocol/client/client_test.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+)
+
+var compressors = map[string]*frame.Codec{
+	"LZ4":    frame.NewCodec(frame.WithCompressor(&compression.Lz4Compressor{})),
+	"SNAPPY": frame.NewCodec(frame.WithCompressor(&compression.SnappyCompressor{})),
+	"NONE":   frame.NewCodec(),
+}
+
+// This test requires a remote server listening on localhost:9042 without authentication.
+func TestRemoteServerNoAuth(t *testing.T) {
```
Comment (adutra) [inline cassandraprotocol/client/client_test.go:20]: These tests will have to be removed because they require an external server, otherwise they can fail. But I included them for now to illustrate how to use this new client-server API.

File: cassandraprotocol/client/client_test.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+					assert.NotNil(t, rows)
+					assert.IsType(t, &message.RowsResult{}, rows.Body.Message)
+					fmt.Printf("CLIENT received: %v\n", rows)
+				})
+			}
+		})
+	}
+}
+
+func TestLocalServer(t *testing.T) {
```
Comment (adutra) [inline cassandraprotocol/client/client_test.go:95]: This test and the following one create a ping-pong game between a client and a server.

File: cassandraprotocol/client/connection.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+func (c *CqlConnection) Send(f *frame.Frame) error {
+	return c.codec.Encode(f, c.conn)
+}
+
+func (c *CqlConnection) Receive() (*frame.Frame, error) {
+	return c.codec.Decode(c.conn)
+}
+
+func (c *CqlConnection) ReceiveHeader() (header *frame.RawHeader, err error) {
```
Comment (adutra) [inline cassandraprotocol/client/connection.go:21]: Example of how to decode only the header and discard the body.

File: cassandraprotocol/client/handshake.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	} else if response, err := client.Receive(); err != nil {
+		return err
+	} else if _, ok := response.Body.Message.(*message.Ready); !ok {
+		return fmt.Errorf("expected READY, got %T", response.Body.Message)
+	} else {
+		return nil
+	}
+}
+
+func HandshakeAuth(client *CqlConnection, version cassandraprotocol.ProtocolVersion, streamId int16, username string, password string) error {
```
Comment (adutra) [inline cassandraprotocol/client/handshake.go:33]: Handshake procedure with authentication enabled. Tested against C* 3.11 and DSE 6.0.

File: cassandraprotocol/client/client_test.go (hunk 1 above)
Comment (joao-r-reis) [reply cassandraprotocol/client/client_test.go:20]: maybe add a check on a test flag or env variable so that by default `go test` doesn't run this

Comment (joao-r-reis) [reply cassandraprotocol/client/client_test.go:20]: Same comment for other tests that require an external server

File: cassandraprotocol/client/client_test.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+					fmt.Printf("SERVER sent:     %v\n", msg)
+
+					header, _ = clientConn.ReceiveHeader()
+					assert.NotNil(t, header)
+					fmt.Printf("CLIENT received: %v\n", header)
+				})
+			}
+		})
+	}
+}
```
Comment (joao-r-reis) [inline cassandraprotocol/client/client_test.go:247]: This minimal cql server will enable so many potential test cases, really cool 👍

File: cassandraprotocol/client/handshake.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	authenticator := &PlainTextAuthenticator{
+		username: username,
+		password: password,
+	}
+	if err := client.Send(NewStartupRequest(client, version, streamId)); err != nil {
+		return err
+	} else if response, err := client.Receive(); err != nil {
```
Comment (joao-r-reis) [inline cassandraprotocol/client/handshake.go:40]: I thought we would port most of the `TestClient` functionality to this repo but this client doesn't listen for responses in the background and map them according to the streamid so it's a bit different. Any reason why you didn't add that logic here?

File: cassandraprotocol/frame/encoder.go [hunk 7]
```diff
@@ -10,24 +10,25 @@ import (
 	"io"
 )
 
-const encodedHeaderLength = 9
-
-func (c *codec) Encode(frame *Frame) (*bytes.Buffer, error) {
+func (c *Codec) Encode(frame *Frame, dest io.Writer) error {
```
Comment (joao-r-reis) [inline cassandraprotocol/frame/encoder.go:13]: I was about to ask you about this, I noticed that we weren't using `io.Writer` but  `bytes.Buffer` which made it a bit harder to use but you already changed it 😄 nice

File: cassandraprotocol/frame/frame.go [hunk 8]
```diff
@@ -110,43 +118,29 @@ func NewResponseFrame(
...
-	if body.Warnings != nil && header.Version < cassandraprotocol.ProtocolVersion4 {
-		return nil, errors.New("warnings require protocol version 4 or higher")
-	}
-	return &Frame{header, body}, nil
+		},
+	}, nil
 }
 
-func (f *Frame) String() string {
+func (f Frame) String() string {
```
Comment (joao-r-reis) [inline cassandraprotocol/frame/frame.go:136]: Why this change for these 3 functions?

**Review by joao-r-reis — COMMENTED** [review]
Looks good overall, I added a couple of comments.

I thought that the cql client would have a bit more functionality considering that the one we're using currently listens for responses in the background and sends them to the appropriate channel using the stream id.

File: cassandraprotocol/client/handshake.go (hunk 6 above)
Comment (adutra) [reply cassandraprotocol/client/handshake.go:40]: I didn't see a reason for doing it so far. Just a basic `Send` and a basic `Receive` method does pretty much everything I wanted to do and test. Handling stream ids sounds a bit like we are starting to create a true driver to me, which I didn't want to do for now. 

One thing to notice is that the `Receive` method blocks until it actually receives a frame. It wasn't problematic for me, but we might indeed have to wrap it in a goroutine if we need more complex test scenarios.

That said, I'm not at all against adding more features to this client, this is really just the minimum that allows us to talk to a server.

File: cassandraprotocol/frame/encoder.go (hunk 7 above)
Comment (adutra) [reply cassandraprotocol/frame/encoder.go:13]: Absolutely, using `io.Writer` and `io.Reader` makes it extremely simple to read and write to `net.Conn` since it satisfies both interfaces. And as a bonus, this means that we don't need to allocate any temporary buffer.

File: cassandraprotocol/client/handshake.go (hunk 6 above)
Comment (joao-r-reis) [reply cassandraprotocol/client/handshake.go:40]: I agree that it wouldn't make sense to do that right now but my impression was that we would just move the `TestClient` code here, not create a new client per say.

File: cassandraprotocol/frame/frame.go (hunk 8 above)
Comment (adutra) [reply cassandraprotocol/frame/frame.go:136]: Because I was having nil pointer panics here. I realized that when a receiver is a pointer, the method gets called event when the actual receiver is nil, as in:

```
var frame *Frame
println(frame.String())
```

The code above panics inside `String()`, _not_ outside of it.

File: cassandraprotocol/client/client_test.go (hunk 1 above)
Comment (adutra) [reply cassandraprotocol/client/client_test.go:20]: Very good idea, will do! 👍

File: cassandraprotocol/frame/frame.go (hunk 8 above)
Comment (joao-r-reis) [reply cassandraprotocol/frame/frame.go:136]: Are you sure that example works? You are passing a pointer but the function receives an actual value. To me the solution would be to check if the receiver is `nil`, no?

File: cassandraprotocol/client/handshake.go (hunk 6 above)
Comment (joao-r-reis) [reply cassandraprotocol/client/handshake.go:40]: I will just make the `TestClient` use this client internally then

File: cassandraprotocol/frame/frame.go (hunk 8 above)
Comment (adutra) [reply cassandraprotocol/frame/frame.go:136]: No it doesn't work either, only your IDE now detects that you are going to have a nil pointer at the call site, not inside the method. But I will revert this change, it is not really related to this PR.

Comment (joao-r-reis) [reply cassandraprotocol/frame/frame.go]: Hmm well it's up to the caller then to make sure it is not calling methods/functions on `nils`

**Review by joao-r-reis — APPROVED** [review]

File: cassandraprotocol/client/handshake.go (hunk 6 above)
Comment (adutra) [reply cassandraprotocol/client/handshake.go:40]: > we talked about moving the existing TestClient code here, not create a new one.

Indeed but a lot of that code was doing encoding and decoding so there was a lot of overlap. In the end I moved only the `Authenticator` part (thanks for that btw) and started the rest from scratch.
